test/test.js

```diff
@@ -131,8 +136,114 @@ describe('Service', function() {
       assert.notExists(properties.getProperty(key));
     });
   });
+
+  describe('#hasAccess()', function() {
+    it('should use the lock to prevent concurrent access', function(done) {
+      var token = {
+        granted_time: 100,
+        expires_in: 100,
+        refresh_token: 'bar'
+      };
+      var properties = new MockProperties();
+      properties.setProperty('oauth2.test', JSON.stringify(token));
+
+      mocks.UrlFetchApp.delayFunction = () => 100;
+      mocks.UrlFetchApp.resultFunction = () =>
+        JSON.stringify({
+          access_token: Math.random().toString(36)
+        });
+
+      var getAccessToken = function() {
+        var service = OAuth2.createService('test')
+            .setClientId('abc')
+            .setClientSecret('def')
+            .setTokenUrl('http://www.example.com')
+            .setPropertyStore(properties)
+            .setLock(new MockLock());
+        if (service.hasAccess()) {
+          return service.getAccessToken();
+        } else {
+          throw new Error('No access: ' + service.getLastError());
+        };
+      }.future();
+
+      Future.task(function() {
```

Comment:
Generators are supported natively in Node/browsers.

Demo: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*

It looks like `fibers/future` was a polyfill used while generators weren't supported. It took me a bit of time to understand the API. Can we try to use native generators rather than Future?

Reply:
Frankly I had a hard enough time getting this all to work as it is, so I'd prefer not to rewrite it all strictly to reduce my reliance on the library. My understanding is that Futures are more than just generators, but I'm not certain.